test-tool/test_async_abort.c

```diff
@@ -0,0 +1,251 @@
+
+		wr_tasks[state.wr_dispatched] = atask;
+		state.wr_dispatched++;
+		logging(LOG_VERBOSE, "WRITE10 queued: %d (CmdSN=0x%x, ITT=0x%x)",
+			state.wr_dispatched, atask->cmdsn, atask->itt);
+	} while ((state.wr_dispatched < max_ios)
+				&& (atask->cmdsn != sd->iscsi_ctx->maxcmdsn));
+
+	/*
+	 * dispatch half of the out queued write requests - don't wait for any
+	 * responses.
+	 */
+	CU_ASSERT_EQUAL(iscsi_out_queue_length(sd->iscsi_ctx),
+			state.wr_dispatched);
+	logging(LOG_VERBOSE, "dispatching %d from out queue...",
+		state.wr_dispatched - (state.wr_dispatched >> 1));
+	while ((uint32_t)iscsi_out_queue_length(sd->iscsi_ctx)
+						> (state.wr_dispatched >> 1)) {
+		struct pollfd pfd;
+
+		pfd.fd = iscsi_get_fd(sd->iscsi_ctx);
+		pfd.events = POLLOUT;	/* only send */
+
+		ret = poll(&pfd, 1, 1000);
+		CU_ASSERT_NOT_EQUAL(ret, -1);
+
+		ret = iscsi_service(sd->iscsi_ctx, pfd.revents);
+		CU_ASSERT_EQUAL(ret, 0);
+	}
+	logging(LOG_VERBOSE, "%d dispatched",
+		state.wr_dispatched - (state.wr_dispatched >> 1));
+
+	/*
+	 * queue aborts - need to use iscsi_task_mgmt_async() instead of
+	 * iscsi_task_mgmt_abort_task_async(), which cancels locally queued
+	 * PDUs.
+	 */
+	for (i = 0; i < state.wr_dispatched; i++) {
+		atask = wr_tasks[i];
+
+		/* need to set private again - it gets stomped */
+		scsi_set_task_private_ptr(atask, &state);
+
+		ret = iscsi_task_mgmt_async(sd->iscsi_ctx,
+					    atask->lun, ISCSI_TM_ABORT_TASK,
+					    atask->itt, atask->cmdsn,
+					    test_async_abort_cb, atask);
+		CU_ASSERT_EQUAL(ret, 0);
+
+		state.abort_dispatched++;
+		logging(LOG_VERBOSE, "ABORT queued: %d of %d (RefCmdSN=0x%x, "
+			"RefITT=0x%x)", state.abort_dispatched,
+			state.wr_dispatched, atask->cmdsn, atask->itt);
+	}
+
+	/* wait for all responses, timeout in 5 seconds */
+	logging(LOG_VERBOSE, "dispatching remaining and handling responses...");
+	timeout_sec = get_clock_sec() + 5;
+	while (get_clock_sec() <= timeout_sec) {
+		struct pollfd pfd;
+
+		pfd.fd = iscsi_get_fd(sd->iscsi_ctx);
+		pfd.events = iscsi_which_events(sd->iscsi_ctx);
+
+		ret = poll(&pfd, 1, 1000);
+		CU_ASSERT_NOT_EQUAL(ret, -1);
+
+		ret = iscsi_service(sd->iscsi_ctx, pfd.revents);
+		CU_ASSERT_EQUAL(ret, 0);
+
+		if ((state.abort_ok + state.wr_good == state.wr_dispatched)
+		 && (state.abort_bad_itt == state.wr_good)) {
+			logging(LOG_VERBOSE, "received all expected responses");
+			break;
+		}
+	}
+
+	logging(LOG_VERBOSE, "%d IOs dispatched, %d IOs completed, "
+		"%d aborts successful, %d aborts unsuccessful",
+		state.wr_dispatched, state.wr_good,
+		state.abort_ok, state.abort_bad_itt);
+
+	for (i = 0; i < state.wr_dispatched; i++) {
+		scsi_free_scsi_task(wr_tasks[i]);
+	}
+
+	/* any writes that were successful should have failed aborts */
+	CU_ASSERT_EQUAL(state.wr_good, state.abort_bad_itt);
+
+	/*
+	 * any successful aborts shouldn't have corresponding write completions
+	 */
+	CU_ASSERT_EQUAL(state.abort_ok, state.wr_dispatched - state.wr_good);
```

Comment:
Actually, these assertions might not quite match rfc3720 -
6.9.  SCSI Timeouts
...
   If the abort request is received and the original command is missing,
   targets MUST consider the original command with that RefCmdSN to be
   received and issue a Task Management response with the response code:
   "Function Complete".  This response concludes the task on both ends.
   If the abort request is received and the target can determine (based
   on the Referenced Task Tag) that the command was received and
   executed and also that the response was sent prior to the abort, then
   the target MUST respond with the response code of "Task Does Not
   Exist".

So not all state.abort_ok responses imply that a task has been aborted at the target - it may also mean that the abort was handled without knowledge that RefCmdSN was handled previously.

I'll withdraw this test for now, and rework it to track all WRITE10 *and* TMF requests individually.